shared/js/background/load.es6.js

```diff
@@ -27,35 +38,65 @@ function returnResponse (xhr, returnType) {
  *  - source: requests are internal by default. set source to 'external' for non-extension URLs
  *  - etag: set an if-none-match header
  */
-function loadExtensionFile (params, cb) {
-    var xhr = new XMLHttpRequest()
+function loadExtensionFile (params) {
+    let xhr = new XMLHttpRequest()
+    let url = params.url
 
     if (params.source === 'external') {
-        xhr.open('GET', params.url)
+        if (dev) {
+            if (url.indexOf('?') > -1) {
+                url += '&'
+            } else {
+                url += '?'
+            }
+
+            url += 'dev=1'
+        }
+
+        xhr.open('GET', url)
+
         if (params.etag) {
             xhr.setRequestHeader('If-None-Match', params.etag)
         }
     } else {
         // set type xhr type tag. Safari internal xhr requests
         // don't set a 200 status so we'll check this type
         xhr.type = 'internal'
-        xhr.open('GET', browserWrapper.getExtensionURL(params.url))
+        xhr.open('GET', browserWrapper.getExtensionURL(url))
     }
 
+    xhr.timeout = 20000
+
     xhr.send(null)
 
-    xhr.onreadystatechange = function () {
-        let done = XMLHttpRequest.DONE ? XMLHttpRequest.DONE : 4
-        if (xhr.readyState === done) {
-            if (xhr.status === 200 || (xhr.type && xhr.type === 'internal')) {
-                cb(returnResponse(xhr, params.returnType), xhr)
+    return new Promise((resolve, reject) => {
+        xhr.ontimeout = () => {
+            reject(new Error(`${url} timed out`))
+        }
```

Comment:
@jdorweiler @kdzwinel I've added the timeout protection here, since ATB can possibly also be affected. I've set a 20s timeout limit, which seems reasonable (e.g. we don't want to wait too long before loading HTTPS rules from the DB)

That said, it seems from some of the answers in https://stackoverflow.com/questions/9502410/ajax-xmlhttprequest-timeout-length-by-browser#9503384 that browsers vary a bit in how they handle XHR timeouts, so hopefully this'll work.

Reply:
Could we do a default of 20 but also pass a value to increase it if we need to?